### PR TITLE
[TASK] Keep track of link during TS rendering

### PR DESCRIPTION
**Pull request by dfeyer**
This change add a new service in the TypoScript package to add dynamic tag
to the current TS rendering path.

This new service is used in the Neos ConvertUrisImplementation to keep
track of link to node and asset and tag the current patch correctly.

Resolves: NEOS-1640

### CODE REVIEW

**Comment by dfeyer** [comment]
:+1:

**Comment by skurfuerst** [comment]
Hey Dominique,

Instead of adding another Service which does this, I'd rather make that part of the TypoScript runtime. Something like an "addTag" method which can be called....

For me, it is a little strange to have a service which contains state, and which is so tightly interleaved with the rendering.

All the best,
Sebastian

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php [hunk 1]
```diff
@@ -144,6 +144,20 @@ public function injectSettings(array $settings)
     }
 
     /**
+     * Add a tag to the current cache segment
+     *
+     * @param string $key
+     * @param string $value
+     */
+    public function addTag($key, $value)
```
Comment (skurfuerst) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php]: what about calling it "addCacheTag"?

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php [hunk 2]
```diff
@@ -144,6 +144,20 @@ public function injectSettings(array $settings)
...
+    {
+        if (isset($this->settings['enableContentCache'])) {
```
Comment (skurfuerst) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php]: huh, why is this check in this direction? "early return if content cache is enabled"?

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php [hunk 3]
```diff
@@ -65,6 +71,31 @@ public function __construct(Runtime $runtime)
...
+        $tag = Functions::ucfirst($key) . 'DynamicTag_' . $value;
+        $this->tags[$tag] = true;
+    }
+
+    /**
+     * @return array
+     */
+    protected function flushTag()
+    {
+        $tags = array_keys(array_filter($this->tags));
```
Comment (skurfuerst) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php]: a small comment would be good what array_filter is supposed to do (strip FALSE values; but how can they happen?)

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php [hunk 4]
```diff
@@ -65,6 +71,31 @@ public function __construct(Runtime $runtime)
...
     /**
+     * @param string $key
+     * @param string $value
+     */
+    public function addTag($key, $value)
+    {
+        $key = trim($key);
+        $value = trim($value);
+        if ($key === '' || $value === '') {
+            return;
```
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php]: not sure if that's a good idea, it's basically ignoring incorrect usage of the method which can be difficult to understand/debug.. maybe an exception is more fitting or what's the thinking behind it?

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php [hunk 5]
```diff
@@ -65,6 +71,31 @@ public function __construct(Runtime $runtime)
     }
 
     /**
+     * @param string $key
+     * @param string $value
```
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php]: @return void

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php [hunk 6]
```diff
@@ -144,6 +144,20 @@ public function injectSettings(array $settings)
     }
 
     /**
+     * Add a tag to the current cache segment
+     *
+     * @param string $key
+     * @param string $value
```
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php]: @return void

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php (hunk 1 above)
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php]: +1

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php [hunk 7]
```diff
@@ -65,6 +71,31 @@ public function __construct(Runtime $runtime)
...
+            return;
+        }
+        $tag = Functions::ucfirst($key) . 'DynamicTag_' . $value;
+        $this->tags[$tag] = true;
+    }
+
+    /**
+     * @return array
+     */
+    protected function flushTag()
```
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php]: ``flushTags``?

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php (hunk 2 above)
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Runtime.php]: yep also using the setting is not so nice, ``if ($this->runtimeContentCache->getEnableContentCache() === false)`` should do the trick

File: TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php [hunk 8]
```diff
@@ -280,10 +311,13 @@ protected function buildCacheTags($configuration, $typoScriptPath, $tsObject)
                     $cacheTags[] = $tagValue;
                 }
             }
+            foreach ($this->flushTag() as $tagKey => $tagValue) {
```
Comment (aertmann) [inline TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/RuntimeContentCache.php]: ``flushTags``?

**Comment by dfeyer** [comment]
New push based on comment from @aertmann and @skurfuerst 

@skurfuerst I removed the array filter, during the first draft I have the idea to introduce a removeCacheTag method, but it's not that easy, not that useful currently, so this array_filter can be removed in flushTags methods.

This version should be OK to merge from my POV :+1:

**Comment by aertmann** [comment]
thanks for taking care of the comments, looks good now :+1: by reading

**Comment by skurfuerst** [comment]
:+1: by reading!

**Pull request merged**